File: long_questions/DSA/Leetcode/golang/8_Binary_Search_on_Answer_Space/875_Koko_Eating_Bananas.go
package main

import (
	"fmt"
	"math"
)

// 875. Koko Eating Bananas
// Time: O(N log M), Space: O(1) where M is max(piles)
func minEatingSpeed(piles []int, h int) int {
	left, right := 1, maxPiles(piles)
	result := right
	
	for left <= right {
		mid := left + (right-left)/2
		
		if canEatAll(piles, h, mid) {
			result = mid
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	
	return result
}

func canEatAll(piles []int, h, speed int) bool {
	hours := 0
	for _, pile := range piles {
		hours += int(math.Ceil(float64(pile) / float64(speed)))
		if hours > h {
			return false
		}
	}
	return hours <= h
}

func maxPiles(piles []int) int {
	max := 0
	for _, pile := range piles {
		if pile > max {
			max = pile
		}
	}
	return max
}

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Binary Search on Answer Space
- **Search Space**: Minimum possible eating speed to maximum possible eating speed
- **Monotonic Property**: If speed X works, any speed > X also works
- **Feasibility Check**: Test if given speed allows eating all bananas in H hours
- **Binary Search**: Narrow down to minimum feasible eating speed

## 2. PROBLEM CHARACTERISTICS
- **Optimization Problem**: Find minimum eating speed that satisfies time constraint
- **Monotonic Feasibility**: Higher speed always makes eating faster
- **Time Constraint**: Must finish all bananas within H hours
- **Speed Bounds**: Between 1 and maximum pile size

## 3. SIMILAR PROBLEMS
- Capacity To Ship Packages Within D Days (LeetCode 1011) - Same pattern
- Split Array Largest Sum (LeetCode 410) - Binary search on max sum
- Minimum Number of Days to Make m Bouquets (LeetCode 1482) - Binary search on days
- Minimum Time to Complete Trips (LeetCode 2187) - Binary search on time

## 4. KEY OBSERVATIONS
- **Monotonic Property**: If speed S works, any S' > S also works
- **Search Bounds**: Lower bound = 1, upper bound = max pile size
- **Feasibility Logic**: Calculate hours needed for given speed
- **Ceiling Division**: Need to round up hours per pile
- **Optimal Solution**: Binary search finds minimum feasible speed

## 5. VARIATIONS & EXTENSIONS
- **Variable Hours**: Find minimum hours for given speed
- **Multiple Monkeys**: Multiple monkeys eating simultaneously
- **Different Pile Types**: Different banana types with different constraints
- **Time-based Eating**: Speed varies by time of day

## 6. INTERVIEW INSIGHTS
- Always clarify: "Can Koko eat fractional bananas? What about H < piles length?"
- Edge cases: single pile, H equals number of piles, H equals 1
- Time complexity: O(N log M) where M is max pile size
- Space complexity: O(1) additional space
- Important: Use ceiling division for hours calculation

## 7. COMMON MISTAKES
- Not setting correct search bounds (left = 1, right = max pile)
- Wrong hours calculation (not using ceiling division)
- Off-by-one errors in binary search
- Not handling edge cases properly
- Using integer division instead of ceiling division

## 8. OPTIMIZATION STRATEGIES
- **Binary Search**: O(N log M) time, optimal for this problem
- **Ceiling Division**: Use math.Ceil or (pile + speed - 1) / speed
- **Early Termination**: Stop when hours exceed H
- **Efficient Bounds**: Tighten search space bounds

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like finding the minimum work rate to meet a deadline:**
- You have tasks (banana piles) that take different amounts of time
- You need to finish all tasks within a deadline (H hours)
- You want to find the minimum work rate that meets the deadline
- If a work rate of X tasks/hour works, any higher rate will also work
- This monotonic property means you can binary search for the minimum

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Array of banana piles, hours H
2. **Goal**: Find minimum eating speed (bananas per hour) to finish all bananas in H hours
3. **Constraint**: Can only eat from one pile per hour, must finish pile before moving
4. **Output**: Minimum eating speed

#### Phase 2: Key Insight Recognition
- **"Binary search natural fit"** → Monotonic feasibility property
- **"Ceiling division key"** → Need to round up hours per pile
- **"Search bounds"** → Between 1 and max pile size
- **"Feasibility check"** → Test if speed allows H-hour eating

#### Phase 3: Strategy Development
```
Human thought process:
"I need to find the minimum eating speed.
If speed X works, any higher speed will also work.
This monotonic property means I can use binary search.
I'll search between 1 (minimum possible) and max pile (maximum possible).
For each speed, I'll calculate hours needed using ceiling division."
```

#### Phase 4: Edge Case Handling
- **Single pile**: Speed = ceil(pile/H)
- **H = number of piles**: Speed = max pile
- **H = 1**: Speed = total bananas
- **All piles same size**: Speed = ceil(pile * count / H)

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
piles = [3,6,7,11], H = 8

Human thinking:
"I need to find the minimum eating speed.
Search range: 1 to 11 (max pile)

Test speed 6:
Hours needed = ceil(3/6) + ceil(6/6) + ceil(7/6) + ceil(11/6)
             = 1 + 1 + 2 + 2 = 6 ≤ 8 ✓ Speed 6 works, try smaller

Test speed 4:
Hours needed = ceil(3/4) + ceil(6/4) + ceil(7/4) + ceil(11/4)
             = 1 + 2 + 2 + 3 = 8 ≤ 8 ✓ Speed 4 works, try smaller

Test speed 3:
Hours needed = ceil(3/3) + ceil(6/3) + ceil(7/3) + ceil(11/3)
             = 1 + 2 + 3 + 4 = 10 > 8 ✗ Speed 3 doesn't work, need larger

Test speed 5:
Hours needed = ceil(3/5) + ceil(6/5) + ceil(7/5) + ceil(11/5)
             = 1 + 2 + 2 + 3 = 8 ≤ 8 ✓ Speed 5 works, try smaller

Test speed 4 (already tested): works, but we need minimum between 4 and 5
Since 3 doesn't work and 4 works, minimum speed = 4 ✓"
```

#### Phase 6: Intuition Validation
- **Why binary search works**: Monotonic feasibility property
- **Why ceiling division**: Must round up hours per pile
- **Why search bounds**: Minimum = 1, maximum = max pile
- **Why O(N log M)**: N for hours calculation, log M for binary search

### Common Human Pitfalls & How to Avoid Them
1. **"Why not use linear search?"** → Too slow, binary search is optimal
2. **"Should I use regular division?"** → No, need ceiling division
3. **"What about search bounds?"** → Must be 1 to max pile
4. **"Can I optimize further?"** → Binary search is already optimal

### Real-World Analogy
**Like finding the minimum typing speed to finish work on time:**
- You have documents of different lengths to type
- You need to finish all documents within a deadline
- You want to find the minimum typing speed that meets the deadline
- If you can type at X words/minute and finish on time, any faster speed will also work
- You can test different speeds and find the minimum that works
- This is exactly what binary search does for you

### Human-Readable Pseudocode
```
function minEatingSpeed(piles, H):
    left = 1           // Minimum possible speed
    right = max(piles) // Maximum possible speed
    result = right
    
    while left <= right:
        mid = left + (right - left) / 2
        
        if canEatAll(piles, H, mid):
            result = mid
            right = mid - 1  // Try smaller speed
        else:
            left = mid + 1   // Need faster speed
    
    return result

function canEatAll(piles, H, speed):
    hours = 0
    for pile in piles:
        hours += ceil(pile / speed)
        if hours > H:
            return false
    return hours <= H
```

### Execution Visualization

### Example: piles = [30,11,23,4,20], H = 6
```
Binary Search Process:
Search range: [1, 30] (1 to max pile)

Test speed 15:
Hours needed = ceil(30/15) + ceil(11/15) + ceil(23/15) + ceil(4/15) + ceil(20/15)
             = 2 + 1 + 2 + 1 + 2 = 8 > 6 ✗ Need faster

Test speed 23:
Hours needed = ceil(30/23) + ceil(11/23) + ceil(23/23) + ceil(4/23) + ceil(20/23)
             = 2 + 1 + 1 + 1 + 1 = 6 ≤ 6 ✓ Try smaller

Test speed 12:
Hours needed = ceil(30/12) + ceil(11/12) + ceil(23/12) + ceil(4/12) + ceil(20/12)
             = 3 + 1 + 2 + 1 + 2 = 9 > 6 ✗ Need faster

Test speed 18:
Hours needed = ceil(30/18) + ceil(11/18) + ceil(23/18) + ceil(4/18) + ceil(20/18)
             = 2 + 1 + 2 + 1 + 2 = 8 > 6 ✗ Need faster

Test speed 21:
Hours needed = ceil(30/21) + ceil(11/21) + ceil(23/21) + ceil(4/21) + ceil(20/21)
             = 2 + 1 + 2 + 1 + 1 = 7 > 6 ✗ Need faster

Minimum speed = 23 ✓
```

### Key Visualization Points:
- **Monotonic property**: Higher speed always works if lower works
- **Ceiling division**: Must round up hours per pile
- **Binary search**: Narrow down to minimum feasible speed
- **Search bounds**: Between 1 and max pile size

### Memory Layout Visualization:
```
Speed Test Visualization:
speed = 23, piles = [30,11,23,4,20]

Hours calculation:
pile 30: ceil(30/23) = 2 hours
pile 11: ceil(11/23) = 1 hour
pile 23: ceil(23/23) = 1 hour
pile 4:  ceil(4/23) = 1 hour
pile 20: ceil(20/23) = 1 hour

Total hours = 6 ≤ H ✓ Speed 23 works!
```

### Time Complexity Breakdown:
- **Binary Search**: O(log M) iterations where M = max(piles)
- **Hours Calculation**: O(N) time per iteration
- **Total Time**: O(N log M)
- **Space Complexity**: O(1) additional space

### Alternative Approaches:

#### 1. Linear Search (O(N × M) time, O(1) space)
```go
func minEatingSpeedLinear(piles []int, h int) int {
    for speed := 1; speed <= maxPiles(piles); speed++ {
        if canEatAll(piles, h, speed) {
            return speed
        }
    }
    return maxPiles(piles)
}
```
- **Pros**: Simple to understand
- **Cons**: Too slow for large inputs

#### 2. Priority Queue Approach (O(N log N) time, O(N) space)
```go
func minEatingSpeedPQ(piles []int, h int) int {
    // This approach doesn't directly solve the problem
    // but could be used for related scheduling problems
    return -1
}
```
- **Pros**: Useful for some scheduling variants
- **Cons**: Not applicable to this specific problem

#### 3. Mathematical Approach (O(N) time, O(1) space)
```go
func minEatingSpeedMath(piles []int, h int) int {
    // For special cases, can use mathematical formulas
    // but general case requires binary search
    return -1
}
```
- **Pros**: Fast for special cases
- **Cons**: Not generally applicable

### Extensions for Interviews:
- **Multiple Monkeys**: Multiple monkeys eating simultaneously
- **Variable Speed**: Speed changes based on time or pile size
- **Different Constraints**: Add constraints on eating patterns
- **Time-based Scheduling**: Speed varies by time of day
- **Resource Allocation**: Generalize to other resource allocation problems
*/
func main() {
	// Test cases
	testCases := []struct {
		piles []int
		h     int
	}{
		{[]int{3, 6, 7, 11}, 8},
		{[]int{30, 11, 23, 4, 20}, 5},
		{[]int{30, 11, 23, 4, 20}, 6},
		{[]int{1}, 1},
		{[]int{100}, 1},
		{[]int{100}, 100},
		{[]int{312884470}, 312884469},
		{[]int{1, 1, 1, 1, 1, 1, 1, 1, 1}, 10},
		{[]int{5, 10, 15, 20, 25}, 15},
	}
	
	for i, tc := range testCases {
		result := minEatingSpeed(tc.piles, tc.h)
		fmt.Printf("Test Case %d: piles=%v, h=%d -> Min speed: %d\n", 
			i+1, tc.piles, tc.h, result)
	}
}

File: long_questions/DSA/Leetcode/java/8_Binary_Search_on_Answer_Space/875_Koko_Eating_Bananas.java
public class KokoEatingBananas {
    
    // 875. Koko Eating Bananas
    // Time: O(N log M), Space: O(1) where M is max(piles)
    public int minEatingSpeed(int[] piles, int h) {
        int left = 1;
        int right = maxPiles(piles);
        int result = right;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (canEatAll(piles, h, mid)) {
                result = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        
        return result;
    }
    
    private boolean canEatAll(int[] piles, int h, int speed) {
        int hours = 0;
        for (int pile : piles) {
            hours += (pile + speed - 1) / speed; // Equivalent to Math.ceil(pile / speed)
            if (hours > h) {
                return false;
            }
        }
        return hours <= h;
    }
    
    private int maxPiles(int[] piles) {
        int max = 0;
        for (int pile : piles) {
            if (pile > max) {
                max = pile;
            }
        }
        return max;
    }
    
    // Alternative approach with detailed explanation
    public int minEatingSpeedDetailed(int[] piles, int h) {
        System.out.println("=== Binary Search on Answer Space ===");
        System.out.printf("Piles: %s, Hours: %d\n", java.util.Arrays.toString(piles), h);
        
        int left = 1;
        int right = maxPiles(piles);
        int result = right;
        
        System.out.printf("Search space: [%d, %d]\n", left, right);
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            System.out.printf("Trying speed: %d bananas/hour\n", mid);
            
            int hoursNeeded = calculateHoursNeeded(piles, mid);
            System.out.printf("  Hours needed: %d\n", hoursNeeded);
            
            if (hoursNeeded <= h) {
                result = mid;
                right = mid - 1;
                System.out.printf("  ✓ Can finish in time, searching lower\n");
            } else {
                left = mid + 1;
                System.out.printf("  ✗ Cannot finish, need higher speed\n");
            }
        }
        
        System.out.printf("Final result: %d bananas/hour\n", result);
        return result;
    }
    
    private int calculateHoursNeeded(int[] piles, int speed) {
        int hours = 0;
        for (int pile : piles) {
            int hoursForPile = (pile + speed - 1) / speed;
            hours += hoursForPile;
            System.out.printf("    Pile %d: %d hours (%.1f hours)\n", 
                pile, hoursForPile, (double) pile / speed);
        }
        return hours;
    }
    
    // Brute force approach for comparison
    public int minEatingSpeedBruteForce(int[] piles, int h) {
        for (int speed = 1; speed <= maxPiles(piles); speed++) {
            if (canEatAll(piles, h, speed)) {
                return speed;
            }
        }
        return maxPiles(piles);
    }
    
    // Optimized version with early termination
    public int minEatingSpeedOptimized(int[] piles, int h) {
        int left = 1;
        int right = maxPiles(piles);
        int result = right;
        
        // Early optimization: if h == piles.length, answer is maxPiles
        if (h == piles.length) {
            return right;
        }
        
        // Early optimization: if h >= sum(piles), answer is 1
        int totalBananas = 0;
        for (int pile : piles) {
            totalBananas += pile;
        }
        if (h >= totalBananas) {
            return 1;
        }
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (canEatAllOptimized(piles, h, mid)) {
                result = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        
        return result;
    }
    
    private boolean canEatAllOptimized(int[] piles, int h, int speed) {
        int hours = 0;
        for (int pile : piles) {
            hours += (pile + speed - 1) / speed;
            
            // Early termination
            if (hours > h) {
                return false;
            }
        }
        return true;
    }
    
    // Version that returns the eating schedule
    public class EatingSchedule {
        int minSpeed;
        java.util.List<java.util.List<Integer>> hourlySchedule;
        int totalHours;
        
        EatingSchedule(int minSpeed, java.util.List<java.util.List<Integer>> hourlySchedule, int totalHours) {
            this.minSpeed = minSpeed;
            this.hourlySchedule = hourlySchedule;
            this.totalHours = totalHours;
        }
    }
    
    public EatingSchedule getEatingSchedule(int[] piles, int h) {
        int minSpeed = minEatingSpeed(piles, h);
        java.util.List<java.util.List<Integer>> schedule = new java.util.ArrayList<>();
        int totalHours = 0;
        
        for (int pile : piles) {
            int remaining = pile;
            java.util.List<Integer> hourlyIntake = new java.util.ArrayList<>();
            
            while (remaining > 0) {
                int eat = Math.min(minSpeed, remaining);
                hourlyIntake.add(eat);
                remaining -= eat;
                totalHours++;
            }
            
            schedule.add(hourlyIntake);
        }
        
        return new EatingSchedule(minSpeed, schedule, totalHours);
    }
    
    // Mathematical approach to find lower bound
    public int minEatingSpeedMathematical(int[] piles, int h) {
        int totalBananas = 0;
        int maxPile = 0;
        
        for (int pile : piles) {
            totalBananas += pile;
            maxPile = Math.max(maxPile, pile);
        }
        
        // Lower bound: ceil(totalBananas / h)
        int lowerBound = (totalBananas + h - 1) / h;
        int upperBound = maxPile;
        
        System.out.printf("Mathematical bounds: [%d, %d]\n", lowerBound, upperBound);
        
        int left = Math.max(1, lowerBound);
        int right = upperBound;
        int result = right;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (canEatAll(piles, h, mid)) {
                result = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        
        return result;
    }
    
    public static void main(String[] args) {
        KokoEatingBananas koko = new KokoEatingBananas();
        
        // Test cases
        int[][][] testCases = {
            {{3, 6, 7, 11}, {8}},
            {{30, 11, 23, 4, 20}, {5}},
            {{30, 11, 23, 4, 20}, {6}},
            {{1}, {1}},
            {{100}, {1}},
            {{100}, {100}},
            {{312884470}, {312884469}},
            {{1, 1, 1, 1, 1, 1, 1, 1, 1}, {10}},
            {{5, 10, 15, 20, 25}, {15}}
        };
        
        String[] descriptions = {
            "Standard case",
            "Tight deadline",
            "Relaxed deadline",
            "Single banana",
            "Large pile, tight deadline",
            "Large pile, relaxed deadline",
            "Very large pile",
            "Many small piles",
            "Increasing piles"
        };
        
        for (int i = 0; i < testCases.length; i++) {
            System.out.printf("Test Case %d: %s\n", i + 1, descriptions[i]);
            int[] piles = testCases[i][0];
            int h = testCases[i][1][0];
            
            int result1 = koko.minEatingSpeed(piles, h);
            int result2 = koko.minEatingSpeedBruteForce(piles, h);
            int result3 = koko.minEatingSpeedOptimized(piles, h);
            int result4 = koko.minEatingSpeedMathematical(piles, h);
            
            System.out.printf("  Binary Search: %d\n", result1);
            System.out.printf("  Brute Force: %d\n", result2);
            System.out.printf("  Optimized: %d\n", result3);
            System.out.printf("  Mathematical: %d\n", result4);
            
            // Get eating schedule
            EatingSchedule schedule = koko.getEatingSchedule(piles, h);
            System.out.printf("  Eating Schedule (speed %d):\n", schedule.minSpeed);
            for (int j = 0; j < schedule.hourlySchedule.size(); j++) {
                System.out.printf("    Pile %d: %s\n", j + 1, schedule.hourlySchedule.get(j));
            }
            System.out.printf("  Total hours: %d\n", schedule.totalHours);
            System.out.println();
        }
        
        // Detailed explanation for one case
        System.out.println("=== Detailed Explanation ===");
        int[] detailedPiles = {30, 11, 23, 4, 20};
        int detailedHours = 6;
        koko.minEatingSpeedDetailed(detailedPiles, detailedHours);
        
        // Performance comparison
        System.out.println("\n=== Performance Comparison ===");
        
        // Large test case
        int[] largePiles = new int[1000];
        for (int i = 0; i < 1000; i++) {
            largePiles[i] = (i % 100) + 1;
        }
        
        long startTime = System.nanoTime();
        int largeResult = koko.minEatingSpeed(largePiles, 500);
        long endTime = System.nanoTime();
        
        System.out.printf("Large test (1000 piles, 500 hours): %d (took %d ns)\n", 
            largeResult, endTime - startTime);
        
        startTime = System.nanoTime();
        int bruteResult = koko.minEatingSpeedBruteForce(largePiles, 500);
        endTime = System.nanoTime();
        
        System.out.printf("Brute force result: %d (took %d ns)\n", 
            bruteResult, endTime - startTime);
        
        // Edge cases
        System.out.println("\n=== Edge Cases ===");
        
        // Minimum possible speed
        int[] minSpeedCase = {1, 1, 1, 1, 1};
        System.out.printf("Minimum speed case: %d\n", koko.minEatingSpeed(minSpeedCase, 5));
        
        // Maximum possible speed
        int[] maxSpeedCase = {1000000, 1000000, 1000000};
        System.out.printf("Maximum speed case: %d\n", koko.minEatingSpeed(maxSpeedCase, 3));
        
        // Very tight deadline
        int[] tightDeadline = {100, 200, 300, 400, 500};
        System.out.printf("Tight deadline case: %d\n", koko.minEatingSpeed(tightDeadline, 5));
    }

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Binary Search on Answer Space
- **Binary Search**: Search for minimum eating speed
- **Feasibility Check**: Can Koko eat all bananas within H hours
- **Search Space**: [1, maxPile] to [sum(piles)]
- **Monotonic**: If speed K works, all larger speeds work

## 2. PROBLEM CHARACTERISTICS
- **Banana Piles**: Array of pile sizes
- **Eating Constraint**: H hours to finish all piles
- **Speed Question**: Find minimum eating speed (bananas/hour)
- **Feasibility**: Check if speed K allows finishing within time limit

## 3. SIMILAR PROBLEMS
- Capacity To Ship Packages Within D Days
- Minimum Number of Days to Make M Bouquets
- Split Array Largest Sum
- Allocate Minimum Number of Books

## 4. KEY OBSERVATIONS
- Binary search works because feasibility is monotonic
- If speed K works, any speed > K also works
- Search space: [1, maxPile] to [sum(piles)]
- Feasibility check: Simulate eating process for given speed

## 5. VARIATIONS & EXTENSIONS
- Return actual eating schedule
- Multiple queries on same piles
- Different eating constraints (e.g., guards watching)
- Real-time streaming version

## 6. INTERVIEW INSIGHTS
- Clarify: "Should I return the eating schedule?"
- Edge cases: empty piles, single pile, H=0
- Time complexity: O(N log M) vs O(NM) brute force
- Space complexity: O(1) vs O(N) for brute force

## 7. COMMON MISTAKES
- Using linear search instead of binary search
- Incorrect feasibility check implementation
- Off-by-one errors in binary search bounds
- Not handling edge cases properly

## 8. OPTIMIZATION STRATEGIES
- Binary search is optimal for single query
- Early termination for obvious cases
- Mathematical bounds for search space
- Pre-sorting if piles are unsorted

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like finding the right eating speed:**
- Koko has banana piles of different sizes
- She eats at a constant rate K (bananas/hour)
- You want to find the minimum K to finish within H hours
- If speed K works, any faster speed also works
- This monotonic property allows binary search

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Array of pile sizes and integer H
2. **Goal**: Find minimum eating speed to finish all piles within H hours
3. **Output**: Minimum feasible speed K

#### Phase 2: Key Insight Recognition
- **"Can I test a speed?"** → Simulate eating process
- **"Is the relationship monotonic?"** → Yes! If speed K works, K+1 works
- **"What's the search range?"** → [1, maxPile] to [sum(piles)]
- **"Can I use binary search?"** → Yes, monotonic feasibility

#### Phase 3: Strategy Development
```
Human thought process:
"I'll use binary search to find the minimum speed:
1. Define search space: [1, maxPile] to [sum(piles)]
2. For each speed mid:
   - Check if Koko can eat all piles within H hours
   - If yes, try smaller speed
   - If no, try larger speed
3. The smallest 'yes' speed is my answer"
```

#### Phase 4: Edge Case Handling
- **Empty piles**: Return 0 (no speed needed)
- **H >= sum(piles)**: Return sum(piles) (eat all at once)
- **Single pile**: Return ceil(pile/H)
- **Large H**: Binary search helps avoid O(NH) brute force

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Starting with: piles = [30,11,23,4,20], H = 5

Human thinking:
"Search space: [1, 30] to [88]

Step 1: mid = (1+30)/2 = 15
Can Koko eat with speed 15?
- Pile 1: ceil(30/15) = 2 hours
- Pile 2: ceil(11/15) = 1 hour
- Pile 3: ceil(23/15) = 2 hours
- Pile 4: ceil(4/15) = 1 hour
- Pile 5: ceil(20/15) = 2 hours
- Total: 2+1+2+1+2 = 8 > 5 hours ✗
- No! Try smaller speed

Step 2: mid = (1+14)/2 = 7
Can Koko eat with speed 7?
- Pile 1: ceil(30/7) = 5 hours
- Pile 2: ceil(11/7) = 2 hours
- Pile 3: ceil(23/7) = 4 hours
- Pile 4: ceil(4/7) = 1 hour
- Pile 5: ceil(20/7) = 3 hours
- Total: 5+2+4+1+3 = 15 > 5 hours ✗
- No! Try smaller speed

Step 3: mid = (1+6)/2 = 3
Can Koko eat with speed 3?
- Pile 1: ceil(30/3) = 10 hours
- Pile 2: ceil(11/3) = 4 hours
- Pile 3: ceil(23/3) = 8 hours
- Pile 4: ceil(4/3) = 2 hours
- Pile 5: ceil(20/3) = 7 hours
- Total: 10+4+8+2+7 = 31 > 5 hours ✗
- No! Try smaller speed

Step 4: mid = (1+2)/2 = 1
Can Koko eat with speed 1?
- Pile 1: ceil(30/1) = 30 hours
- Pile 2: ceil(11/1) = 11 hours
- Pile 3: ceil(23/1) = 23 hours
- Pile 4: ceil(4/1) = 4 hours
- Pile 5: ceil(20/1) = 20 hours
- Total: 30+11+23+4+20 = 88 > 5 hours ✗
- No! Try smaller speed

Wait, I made an error. Let me recalculate:

Actually, with speed 4:
- Pile 1: ceil(30/4) = 8 hours
- Pile 2: ceil(11/4) = 3 hours
- Pile 3: ceil(23/4) = 6 hours
- Pile 4: ceil(4/4) = 1 hour
- Pile 5: ceil(20/4) = 5 hours
- Total: 8+3+6+1+5 = 23 > 5 hours ✗

Wait, let me try speed 2:
- Total: 15+4+8+2+7 = 36 > 5 hours ✗

Actually, let me try speed 3:
- Total: 10+4+8+2+7 = 31 > 5 hours ✗

Actually, let me try speed 5:
- Total: 6+4+8+2+4 = 24 > 5 hours ✗

Actually, let me try speed 6:
- Total: 5+4+8+2+4 = 23 > 5 hours ✗

Actually, let me try speed 7:
- Total: 5+4+8+2+3 = 22 > 5 hours ✗

Actually, let me try speed 8:
- Total: 4+4+8+2+3 = 21 > 5 hours ✗

Actually, let me try speed 9:
- Total: 4+4+8+2+3 = 21 > 5 hours ✗

Actually, let me try speed 10:
- Total: 3+4+8+2+2 = 19 > 5 hours ✗

Actually, let me try speed 11:
- Total: 3+4+8+2+2 = 19 > 5 hours ✗

Actually, let me try speed 12:
- Total: 3+4+8+2+2 = 19 > 5 hours ✗

Actually, let me try speed 13:
- Total: 3+4+8+2+2 = 19 > 5 hours ✗

Actually, let me try speed 14:
- Total: 3+4+8+2+2 = 19 > 5 hours ✗

Actually, let me try speed 15:
- Total: 2+4+8+2+2 = 18 > 5 hours ✗

Actually, let me try speed 22:
- Total: 2+4+8+2+2 = 18 > 5 hours ✗

Actually, let me try speed 23:
- Total: 2+4+8+2+2 = 18 > 5 hours ✗

Actually, let me try speed 30:
- Pile 1: ceil(30/30) = 1 hour
- Pile 2: ceil(11/30) = 1 hour
- Pile 3: ceil(23/30) = 1 hour
- Pile 4: ceil(4/30) = 1 hour
- Pile 5: ceil(20/30) = 1 hour
- Total: 1+1+1+1+1 = 5 ≤ 5 hours ✓
- Yes! Answer is 30"
```

#### Phase 6: Intuition Validation
- **Why it works**: Monotonic feasibility enables binary search
- **Why it's efficient**: O(N log M) vs O(NH) brute force
- **Why it's correct**: Binary search finds minimal feasible speed

### Common Human Pitfalls & How to Avoid Them
1. **"Why not just try all speeds?"** → Too slow O(NH)
2. **"What's wrong with linear search?"** → Misses logarithmic improvement
3. **"How to check feasibility?"** → Must simulate actual eating process
4. **"What about the bounds?"** → Lower bound is 1, upper is sum(piles)

### Real-World Analogy
**Like finding the right machine speed:**
- You have tasks of different sizes (piles)
- You want to complete all tasks within H hours (deadline)
- You test different machine speeds
- If a speed works, any faster speed also works
- You want the slowest speed that meets the deadline
- This is exactly the binary search problem!

### Human-Readable Pseudocode
```
function minEatingSpeed(piles, H):
    left = 1
    right = sum(piles)
    result = right
    
    while left <= right:
        mid = left + (right - left) / 2
        
        if canEatAll(piles, H, mid):
            result = mid
            right = mid - 1
        else:
            left = mid + 1
    
    return result

function canEatAll(piles, H, speed):
    hours = 0
    currentLoad = 0
    
    for pile in piles:
        hours += (pile + speed - 1) / speed
        currentLoad += pile
        
        if hours > H:
            return false
    
    return true
```

### Execution Visualization

### Example: piles = [30,11,23,4,20], H = 5
```
Initial: piles = [30,11,23,4,20], H = 5
Search space: [1, 30] to [88]

Step 1: mid = 15
Testing speed 15:
- Hours: 2+1+2+1+2 = 8 > 5 ✗
Result: Cannot eat, search higher

Step 2: mid = 7
Testing speed 7:
- Hours: 5+2+4+1+3 = 15 > 5 ✗
Result: Cannot eat, search higher

Step 3: mid = 3
Testing speed 3:
- Hours: 10+4+8+2+7 = 31 > 5 ✗
Result: Cannot eat, search higher

Step 4: mid = 1
Testing speed 1:
- Hours: 30+11+23+4+20 = 88 > 5 ✗
Result: Cannot eat, search higher

Wait, I think I made an error. Let me recalculate properly:

Actually, with speed 30:
- Pile 1: ceil(30/30) = 1 hour
- Pile 2: ceil(11/30) = 1 hour
- Pile 3: ceil(23/30) = 1 hour
- Pile 4: ceil(4/30) = 1 hour
- Pile 5: ceil(20/30) = 1 hour
- Total: 1+1+1+1+1 = 5 ≤ 5 hours ✓
- Yes! Answer is 30"
```

### Key Visualization Points:
- **Binary search** on speed values
- **Feasibility check** simulates eating process
- **Monotonic property**: If speed works, larger speeds work
- **Search space reduction**: Narrow down to minimum feasible speed

### Memory Layout Visualization:
```
Piles: [30][11][23][4][20]
Speed 15: [2h][1h][2h][1h][2h]
          Day 1  Day 2  Day 3  Day 4  Day 5
          ✓     ✗     ✗     ✗     ✗     ✗

Speed 30:  [1h][1h][1h][1h][1h]
          Day 1  Day 2  Day 3  Day 4  Day 5
          ✓     ✓     ✓     ✓     ✓     ✓

Final Answer: 30 ✓
```

### Time Complexity Breakdown:
- **Binary Search**: O(log M) iterations where M = sum(piles)
- **Feasibility Check**: O(N) per iteration
- **Total**: O(N log M) time, O(1) space
- **Optimal**: Best possible for single query
- **vs Brute Force**: O(NH) where H is max pile size
*/
